### PR TITLE
(CDAP-2836) Add support for restart specific CDAP system services instances

### PR DESCRIPTION
**Pull request by henry-cask**
Add support for restart for CDAP system services instances. Need to be able to send restart commands to individual instances in System Services for CDAP Master.

Currently, to restart one of the CDAP master services (data-fabric, explore, metrics, etc.), you have to restart them all. It should be possible to just stop/start/restart a specific one.

### CODE REVIEW

**Comment by henry-cask** [comment]
This one pending Apache Twill PR:
https://github.com/apache/incubator-twill/pull/52

File: cdap-common/src/main/java/co/cask/cdap/common/twill/AbstractDistributedMasterServiceManager.java [hunk 1]
```diff
@@ -179,4 +179,41 @@ protected final HttpResponseStatus checkGetStatus(String url) throws Exception {
...
+    return true;
+  }
+
+  @Override
+  public boolean restartInstance(int instanceId, int... moreInstanceIds) {
+    Preconditions.checkArgument(serviceName != null);
+    // Get Twill controller for the service
+    try {
+      Iterable<TwillController> twillControllerList = twillRunnerService.lookup(Constants.Service.MASTER_SERVICES);
+      if (twillControllerList != null) {
```
Comment (chtyim) [inline cdap-common/src/main/java/co/cask/cdap/common/twill/AbstractDistributedMasterServiceManager.java]: I don't think lookup will every return null.

Comment (hsaputra) [inline cdap-common/src/main/java/co/cask/cdap/common/twill/AbstractDistributedMasterServiceManager.java]: Ah cool, will just iterate through it then =)

Comment (henry-cask) [inline cdap-common/src/main/java/co/cask/cdap/common/twill/AbstractDistributedMasterServiceManager.java]: Ah cool, will just iterate through it then =)

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/store/ServiceStore.java [hunk 2]
```diff
@@ -39,4 +39,25 @@
...
+  /**
+   * Update the service instances restart request.
+   *
+   * @param serviceName Service name to be restarted.
+   * @param startTime Start time in Ms from Epoch.
+   * @param endTime End time in Ms from Epoch.
+   * @param isSuccess Whether or not the operation successful.
+   * @param instanceId The instance Id to be restarted. If null then it restart all instances.
+   */
+  void setRestartInstancesRequest(String serviceName, long startTime, long endTime, boolean isSuccess,
```
Comment (jwang47) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/store/ServiceStore.java]: Seems weird to restart all if `instanceId` is null. Why not have two methods, `requestInstanceRestart` and `requestAllInstancesRestart`?

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/DatasetServiceStore.java [hunk 3]
```diff
@@ -64,4 +72,41 @@ protected void startUp() throws Exception {
   protected void shutDown() throws Exception {
     table.close();
   }
+
+  @Override
+  public synchronized void setRestartInstancesRequest(String serviceName, long startTimeMs, long endTimeMs,
+                                                      boolean isSuccess, Integer instanceId) {
```
Comment (jwang47) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/DatasetServiceStore.java]: `@Nullable Integer instanceId`

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/DatasetServiceStore.java [hunk 4]
```diff
@@ -64,4 +72,41 @@ protected void startUp() throws Exception {
...
+    RestartServiceInstancesStatus.RestartStatus status = RestartServiceInstancesStatus.RestartStatus.SUCCESS;
```
Comment (jwang47) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/DatasetServiceStore.java]: Why not use `status = isSuccess ? SUCCESS : FAILURE` instead of setting then setting again conditionally?

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/DatasetServiceStore.java [hunk 5]
```diff
@@ -64,4 +72,41 @@ protected void startUp() throws Exception {
...
+    if (!isSuccess) {
+      status = RestartServiceInstancesStatus.RestartStatus.FAILURE;
+    }
+
+    ImmutableSet.Builder<Integer> instanceIdsBuilder = ImmutableSet.builder();
+    if (instanceId == null) {
+      int instanceCount = (this.getServiceInstance(serviceName) == null) ? 0 : this.getServiceInstance(serviceName);
```
Comment (jwang47) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/DatasetServiceStore.java]: Can alternatively use `Objects.firstNonNull()`.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/MonitorHandler.java [hunk 6]
```diff
@@ -91,4 +91,36 @@ public void monitor(HttpRequest request, HttpResponder responder,
...
+   */
+  @Path("/system/services/{service-name}/instances/restart")
+  @PUT
+  public void restartServiceInstances(HttpRequest request, HttpResponder responder,
+                                      @PathParam("service-name") final String serviceName) {
+    super.restartServiceInstances(request, responder, serviceName);
+  }
+
+  /**
+   * Send request to restart all instances for a CDAP system service.
```
Comment (jwang47) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/MonitorHandler.java]: Fix javadoc

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/store/ServiceStore.java (hunk 2 above)
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/store/ServiceStore.java]: +1

File: cdap-common/src/main/java/co/cask/cdap/common/twill/AbstractDistributedMasterServiceManager.java [hunk 7]
```diff
@@ -179,4 +179,41 @@ protected final HttpResponseStatus checkGetStatus(String url) throws Exception {
       }
     }
   }
+
+  @Override
+  public boolean restartAllInstances() {
+    Preconditions.checkArgument(serviceName != null);
```
Comment (jwang47) [inline cdap-common/src/main/java/co/cask/cdap/common/twill/AbstractDistributedMasterServiceManager.java]: Better to have a slightly more descriptive message here. E.g. "Name of service to restart is missing"

File: cdap-common/src/main/java/co/cask/cdap/common/twill/AbstractDistributedMasterServiceManager.java [hunk 8]
```diff
@@ -179,4 +179,41 @@ protected final HttpResponseStatus checkGetStatus(String url) throws Exception {
...
+      }
+    } catch (Throwable t) {
+      LOG.error("Could not change service instance of {} : {}", serviceName, t.getMessage(), t);
+      return false;
+    }
+    return true;
+  }
+
+  @Override
+  public boolean restartInstance(int instanceId, int... moreInstanceIds) {
```
Comment (jwang47) [inline cdap-common/src/main/java/co/cask/cdap/common/twill/AbstractDistributedMasterServiceManager.java]: Why return boolean instead of throwing exception?

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java [hunk 9]
```diff
@@ -202,4 +202,72 @@ private int getSystemServiceInstanceCount(String serviceName) throws Exception {
...
+        return;
+      }
+
+      // Need to get Twill controller for the service.
+      MasterServiceManager masterServiceManager = serviceManagementMap.get(serviceName);
+      if (masterServiceManager.restartAllInstances()) {
+        responder.sendStatus(HttpResponseStatus.OK);
+      } else {
+        isSuccess = false;
+        responder.sendString(HttpResponseStatus.BAD_REQUEST, "Operation did not succeed");
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java]: it can also fail due to Internal_Server_Error right?

Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java]: api isn't descriptive about the failure case.

Comment (jwang47) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java]: Shouldn't this actually be Internal Server Error rather than Bad Request? Bad Request is more for something wrong with the input/request.

Comment (henry-cask) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java]: We could use Internal Server Error.

File: cdap-common/src/main/java/co/cask/cdap/common/twill/AbstractDistributedMasterServiceManager.java (hunk 8 above)
Comment (henry-cask) [inline cdap-common/src/main/java/co/cask/cdap/common/twill/AbstractDistributedMasterServiceManager.java]: It is consistent with rest of methods in this class.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java [hunk 10]
```diff
@@ -202,4 +202,72 @@ private int getSystemServiceInstanceCount(String serviceName) throws Exception {
...
+      MasterServiceManager masterServiceManager = serviceManagementMap.get(serviceName);
+      if (masterServiceManager.restartInstance(instanceId)) {
+        responder.sendStatus(HttpResponseStatus.OK);
+      } else {
+        isSuccess = false;
+        responder.sendString(HttpResponseStatus.BAD_REQUEST, "Operation did not succeed");
+      }
+
+    } catch (Exception e) {
+      isSuccess = false;
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java]: could be useful to add log here. LOG.Warn ?

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/store/ServiceStore.java (hunk 2 above)
Comment (henry-cask) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/store/ServiceStore.java]: Agree with 2 separate methods

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/MonitorHandler.java (hunk 6 above)
Comment (henry-cask) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/MonitorHandler.java]: Fix which part?

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java [hunk 11]
```diff
@@ -202,4 +202,72 @@ private int getSystemServiceInstanceCount(String serviceName) throws Exception {
...
+                           String.format("Error starting instance %d for service: %s", instanceId, serviceName));
+    } finally {
+      // Record the time
+      long endTimeMs = System.currentTimeMillis();
+      serviceStore.setRestartInstancesRequest(serviceName, startTimeMs, endTimeMs, isSuccess, null);
+    }
+  }
+
+  public void getLatestRestartServiceInstance(HttpRequest request, HttpResponder responder, String serviceName) {
+    Object restartRequest = serviceStore.getLatestRestartInstancesRequest(serviceName);
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java]: why Object? why not RestartServiceInstancesStatus ?

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java [hunk 12]
```diff
@@ -202,4 +202,72 @@ private int getSystemServiceInstanceCount(String serviceName) throws Exception {
...
+    if (restartRequest == null) {
+      responder.sendString(HttpResponseStatus.NOT_FOUND,
+                           String.format("No restart instances request for %s", serviceName));
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java]: ``No restart instances request found for``

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/MonitorHandler.java (hunk 6 above)
Comment (jwang47) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/MonitorHandler.java]: Isn't this method for restarting a single instance?

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/DatasetServiceStore.java [hunk 13]
```diff
@@ -64,4 +72,36 @@ protected void startUp() throws Exception {
...
+
+  @Override
+  public synchronized void setRestartInstancesRequest(String serviceName, long startTimeMs, long endTimeMs,
+                                                      boolean isSuccess, Integer instanceId) {
+    RestartServiceInstancesStatus.RestartStatus status = RestartServiceInstancesStatus.RestartStatus.SUCCESS;
+    if (!isSuccess) {
+      status = RestartServiceInstancesStatus.RestartStatus.FAILURE;
+    }
+
+    ImmutableSet.Builder<Integer> instanceIdsBuilder = ImmutableSet.builder();
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/DatasetServiceStore.java]: why do you need a builder ?  can just have an List<Integer> right ?

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/DatasetServiceStore.java (hunk 4 above)
Comment (henry-cask) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/DatasetServiceStore.java]: agree

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java [hunk 14]
```diff
@@ -202,4 +202,72 @@ private int getSystemServiceInstanceCount(String serviceName) throws Exception {
...
+                           String.format("Error starting instances for service: %s", serviceName));
+    } finally {
+      // Record the time
+      long endTimeMs = System.currentTimeMillis();
+      serviceStore.setRestartInstancesRequest(serviceName, startTimeMs, endTimeMs, isSuccess, null);
+    }
+  }
+
+  public void restartServiceInstance(HttpRequest request, HttpResponder responder, final String serviceName,
+                                     final int instanceId) {
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java]: this method and ``restartServiceInstances`` are almost similar, can you remove duplication by extracting into a common method.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/MonitorHandler.java (hunk 6 above)
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/MonitorHandler.java]: invalid comment. should be ``restart single instance identified by <instance-id>``

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/MonitorHandler.java [hunk 15]
```diff
@@ -91,4 +91,36 @@ public void monitor(HttpRequest request, HttpResponder responder,
...
+  @Path("/system/services/{service-name}/instances/{instance-id}/restart")
+  @PUT
+  public void restartServiceInstance(HttpRequest request, HttpResponder responder,
+                                     @PathParam("service-name") final String serviceName,
+                                     @PathParam("instance-id") final int instanceId) {
+    super.restartServiceInstance(request, responder, serviceName, instanceId);
+  }
+
+  /**
+   * Send request to get the latest restart instances request for a CDAP system service.
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/MonitorHandler.java]: ``get the status of latest restart instances request``

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/MonitorHandlerTest.java [hunk 16]
```diff
@@ -97,4 +99,27 @@ public void testInstances() throws Exception {
...
   }
+
+  @Test
+  public void testRestartInstances() throws Exception {
+    String path = String.format("system/services/%s/instances/restart", Constants.Service.APP_FABRIC_HTTP);
+    HttpURLConnection urlConn = openURL(path, HttpMethod.PUT);
+
+    Assert.assertEquals(HttpResponseStatus.OK.getCode(), urlConn.getResponseCode());
+
+    TimeUnit.SECONDS.sleep(5L);
```
Comment (gsps1) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/MonitorHandlerTest.java]: anyway we can avoid this ?

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/MonitorHandlerTest.java [hunk 17]
```diff
@@ -97,4 +99,27 @@ public void testInstances() throws Exception {
...
+
+    path = String.format("system/services/%s/instances/restart", Constants.Service.APP_FABRIC_HTTP);
```
Comment (gsps1) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/MonitorHandlerTest.java]: i don't see a change in path. remove?

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/MonitorHandlerTest.java [hunk 18]
```diff
@@ -97,4 +99,27 @@ public void testInstances() throws Exception {
...
+    urlConn = openURL(path, HttpMethod.GET);
+
+    Assert.assertEquals(HttpResponseStatus.OK.getCode(), urlConn.getResponseCode());
+
+    RestartServiceInstancesStatus result =
+      GSON.fromJson(new String(ByteStreams.toByteArray(urlConn.getInputStream()), Charsets.UTF_8),
+                    RestartServiceInstancesStatus.class);
```
Comment (gsps1) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/MonitorHandlerTest.java]: we usually recommend using HttpGet, HttpPost api. that will make the decoding easier as well. as you can do something like, ``GSON.fromJson(EntityUtils.toString(response.getEntity(), Charsets.UTF_8), RestartServiceInstancesStatus.class)``

File: cdap-common/src/main/java/co/cask/cdap/common/twill/AbstractDistributedMasterServiceManager.java [hunk 19]
```diff
@@ -179,4 +179,38 @@ protected final HttpResponseStatus checkGetStatus(String url) throws Exception {
       }
     }
   }
+
+  @Override
+  public boolean restartAllInstances() {
+    Preconditions.checkArgument(serviceName != null);
```
Comment (gsps1) [inline cdap-common/src/main/java/co/cask/cdap/common/twill/AbstractDistributedMasterServiceManager.java]: you can use ``Preconditions.checkNotNull``

File: cdap-common/src/main/java/co/cask/cdap/common/twill/AbstractDistributedMasterServiceManager.java [hunk 20]
```diff
@@ -179,4 +179,38 @@ protected final HttpResponseStatus checkGetStatus(String url) throws Exception {
...
+    try {
+      Iterable<TwillController> twillControllerList = twillRunnerService.lookup(Constants.Service.MASTER_SERVICES);
```
Comment (gsps1) [inline cdap-common/src/main/java/co/cask/cdap/common/twill/AbstractDistributedMasterServiceManager.java]: rename to ``twillControllers`` ?

File: cdap-common/src/main/java/co/cask/cdap/common/twill/AbstractDistributedMasterServiceManager.java [hunk 21]
```diff
@@ -179,4 +179,38 @@ protected final HttpResponseStatus checkGetStatus(String url) throws Exception {
...
+      for (TwillController twillController : twillControllerList) {
+        // Call restart instances
+        twillController.restartAllInstances(serviceName).get();
+      }
+
+    } catch (Throwable t) {
+      LOG.error("Could not change service instance of {} : {}", serviceName, t.getMessage(), t);
```
Comment (gsps1) [inline cdap-common/src/main/java/co/cask/cdap/common/twill/AbstractDistributedMasterServiceManager.java]: ``Could not restart``

File: cdap-common/src/main/java/co/cask/cdap/common/twill/MasterServiceManager.java [hunk 22]
```diff
@@ -86,5 +86,20 @@
 
   SystemServiceLiveInfo getLiveInfo();
 
-  //TODO: Add method to get the metrics name to get event rate on UI
```
Comment (gsps1) [inline cdap-common/src/main/java/co/cask/cdap/common/twill/MasterServiceManager.java]: is this to-do irrelevant? why is it removed

File: cdap-proto/src/main/java/co/cask/cdap/proto/RestartServiceInstancesStatus.java [hunk 23]
```diff
@@ -0,0 +1,65 @@
...
+ */
+package co.cask.cdap.proto;
+
+import java.util.Set;
+
+/**
+ * Class to describe status of restart instances of service request.
+ */
+public class RestartServiceInstancesStatus {
+  private final Set<Integer> instanceIds;
```
Comment (gsps1) [inline cdap-proto/src/main/java/co/cask/cdap/proto/RestartServiceInstancesStatus.java]: why this has to be a ``Set`` ?

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/MonitorHandler.java (hunk 6 above)
Comment (henry-cask) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/MonitorHandler.java]: Ah yes, sorry the review did not show which method being commented on.

File: cdap-proto/src/main/java/co/cask/cdap/proto/RestartServiceInstancesStatus.java (hunk 23 above)
Comment (jwang47) [inline cdap-proto/src/main/java/co/cask/cdap/proto/RestartServiceInstancesStatus.java]: Makes sense to me. Doesn't make sense to have multiple instances of the same ID in a restart status.

Comment (gsps1) [inline cdap-proto/src/main/java/co/cask/cdap/proto/RestartServiceInstancesStatus.java]: i see. that makes sense.

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/MonitorHandlerTest.java (hunk 16 above)
Comment (henry-cask) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/MonitorHandlerTest.java]: Yes we can but lets address it in bug fix mode to make sure we make time to feature freeze.

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/MonitorHandlerTest.java (hunk 17 above)
Comment (henry-cask) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/MonitorHandlerTest.java]: One is GET and one is PUT

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/MonitorHandlerTest.java (hunk 18 above)
Comment (henry-cask) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/MonitorHandlerTest.java]: The other method is the test is using GSON so lets keep it that way and address it in separate PR.

File: cdap-proto/src/main/java/co/cask/cdap/proto/RestartServiceInstancesStatus.java (hunk 23 above)
Comment (henry-cask) [inline cdap-proto/src/main/java/co/cask/cdap/proto/RestartServiceInstancesStatus.java]: This is to prevent duplicate.

File: cdap-common/src/main/java/co/cask/cdap/common/twill/MasterServiceManager.java (hunk 22 above)
Comment (henry-cask) [inline cdap-common/src/main/java/co/cask/cdap/common/twill/MasterServiceManager.java]: Ah accidental remove. Will put it back

Comment (henry-cask) [inline cdap-common/src/main/java/co/cask/cdap/common/twill/MasterServiceManager.java]: Ah no, this is due to some mistake I made commit to my local develop branch. Ignore this one =)

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java (hunk 10 above)
Comment (henry-cask) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java]: No log for rest of classes, will add more logging in separate PR.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/DatasetServiceStore.java (hunk 3 above)
Comment (henry-cask) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/DatasetServiceStore.java]: No longer needed with separate method.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java (hunk 14 above)
Comment (henry-cask) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java]: I rename the other as restartAllServiceInstances to separate restart all or an instance for a system.

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/MonitorHandlerTest.java (hunk 16 above)
Comment (jwang47) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/MonitorHandlerTest.java]: +1, please create a JIRA and add a comment linking the JIRA issue though.

Comment (henry-cask) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/MonitorHandlerTest.java]: Issue filed: https://issues.cask.co/browse/CDAP-3004

**Comment by henry-cask** [comment]
Updated the PR to address comments. CC @jwang47, @gsps1

**Comment by jwang47** [comment]
Instead of amending the commit, please add a new commit when addressing comments as it makes it much easier to view the changes that addressed the comments. Then squash just before merging.

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/store/ServiceStore.java [hunk 24]
```diff
@@ -39,4 +40,35 @@
...
+   * Update the service instances restart request.
+   *
+   * @param serviceName Service name to be restarted.
+   * @param startTime Start time in Ms from Epoch.
+   * @param endTime End time in Ms from Epoch.
+   * @param isSuccess Whether or not the operation successful.
+   * @param instanceId The instance Id to be restarted.
+   */
+  void setRestartInstancesRequest(String serviceName, long startTime, long endTime, boolean isSuccess,
+                                  Integer instanceId);
```
Comment (jwang47) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/store/ServiceStore.java]: Should be `int` right? Since it's no longer nullable.

File: cdap-common/src/main/java/co/cask/cdap/common/twill/AbstractDistributedMasterServiceManager.java (hunk 8 above)
Comment (jwang47) [inline cdap-common/src/main/java/co/cask/cdap/common/twill/AbstractDistributedMasterServiceManager.java]: I think this should be changed, but probably in another PR.. seems bad to log exception here instead of propagating so that the caller knows what happened.

Comment (henry-cask) [inline cdap-common/src/main/java/co/cask/cdap/common/twill/AbstractDistributedMasterServiceManager.java]: Yeah, we could send another PR to change all "set" methods to throw exceptions at fault. For now I prefer to keep it consistent.

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/store/ServiceStore.java (hunk 24 above)
Comment (henry-cask) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/store/ServiceStore.java]: Oops yeah, missing commit push

**Comment by henry-cask** [comment]
Will do it next time

**Comment by jwang47** [comment]
LGTM

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/MonitorHandlerTest.java (hunk 17 above)
Comment (gsps1) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/MonitorHandlerTest.java]: that should not change the string right

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java (hunk 14 above)
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java]: oh, i meant the code inside the method.

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/MonitorHandlerTest.java (hunk 18 above)
Comment (gsps1) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/MonitorHandlerTest.java]: we will still use GSON, instead of urlConnection we will use HttpPost and HttpGet. but its fine if you don't want to change it.

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/MonitorHandlerTest.java (hunk 17 above)
Comment (henry-cask) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/MonitorHandlerTest.java]: Ah you mean the path? Will update the tests

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/MonitorHandlerTest.java (hunk 16 above)
Comment (henry-cask) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/MonitorHandlerTest.java]: Actually we dont need sleep here, we just need to verify the request is successfully submitted.

**Comment by henry-cask** [comment]
I have updated the PR with some changes @jwang47, @gsps1, could you help do another pass?

**Comment by jwang47** [comment]
LGTM

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java (hunk 14 above)
Comment (henry-cask) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java]: Ah got it, will converge into one helper method.

**Comment by henry-cask** [comment]
@gsps1 updated the PR to address your concern

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java [hunk 25]
```diff
@@ -202,4 +202,51 @@ private int getSystemServiceInstanceCount(String serviceName) throws Exception {
...
+
+      MasterServiceManager masterServiceManager = serviceManagementMap.get(serviceName);
+      isSuccess = (instanceId < 0) ? masterServiceManager.restartAllInstances() :
+        masterServiceManager.restartInstance(instanceId);
+
+      if (isSuccess) {
+        responder.sendStatus(HttpResponseStatus.OK);
+      } else {
+        // throw exception to have unified error handling.
+        throw new RuntimeException();
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java]: have some message.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java [hunk 26]
```diff
@@ -202,4 +202,51 @@ private int getSystemServiceInstanceCount(String serviceName) throws Exception {
...
+      }
+    } catch (Exception e) {
+      isSuccess = false;
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java]: log the exception, so its useful for debug

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java (hunk 25 above)
Comment (jwang47) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java]: `masterServiceManager.restartAllInstances()` already logs, so I don't think it's necessary here. This does need to change though, probably in another PR.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java [hunk 27]
```diff
@@ -202,4 +202,51 @@ private int getSystemServiceInstanceCount(String serviceName) throws Exception {
...
+      responder.sendString(HttpResponseStatus.INTERNAL_SERVER_ERROR,
+                           String.format("Error restarting instance %d for service: %s", instanceId, serviceName));
+    } finally {
+      // Record the time
+      long endTimeMs = System.currentTimeMillis();
+      serviceStore.setRestartInstanceRequest(serviceName, startTimeMs, endTimeMs, isSuccess, instanceId);
+    }
+  }
+
+  public void getLatestRestartServiceInstance(HttpRequest request, HttpResponder responder, String serviceName) {
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java]: status? can you make the api consistent

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java (hunk 25 above)
Comment (henry-cask) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java]: Since the exception only used to unified error messages in catch clause

**Comment by gsps1** [comment]
just minor nits. LGTM otherwise

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java (hunk 27 above)
Comment (henry-cask) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java]: will do

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java (hunk 25 above)
Comment (henry-cask) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java]: Lets just fix it then, I will change the underlying service manager to throw exception rather than returning boolean.

**Comment by henry-cask** [comment]
Thanks for the review, if no more comments, pending successful build will rebase commits into one.

**Comment by henry-cask** [comment]
Waiting for successful build ...

**Comment by gsps1** [comment]
http://builds.cask.co/browse/CDAP-DUT2204-11

**Comment by gsps1** [comment]
build green - merging

**Comment by henry-cask** [comment]
Successful build: 
http://builds.cask.co/browse/CDAP-DUT2204-11
